// @vercel/postgres-kysely/dist/index.js
// src/index.ts
import { PostgresDialect, PostgresDriver, Kysely } from "kysely";
import { createPool } from "@vercel/postgres";

// src/error.ts
var VercelPostgresKyselyError = class extends Error {
  constructor(code, message) {
    super(`VercelPostgresError - '${code}': ${message}`);
    this.code = code;
    this.name = "VercelPostgresError";
  }
};

// src/index.ts
var VercelPostgresDialect = class extends PostgresDialect {
  constructor(config) {
    super(config);
    this.config = config;
  }
  createDriver() {
    return new VercelPostgresPoolDriver(this.config);
  }
};
var VercelPostgresPoolDriver = class extends PostgresDriver {
  // Rather than trying to rebuild a perfectly good connection pool,
  // we can just use a proxy to throw if the user tries to stream.
  async acquireConnection() {
    const connection = await super.acquireConnection();
    return new Proxy(connection, {
      // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
      get(target, p) {
        const original = target[p];
        if (p === "streamQuery" && typeof original === "function") {
          throw new VercelPostgresKyselyError(
            "kysely_streaming_not_supported",
            "Streaming is not supported yet."
          );
        }
        if (typeof original === "function") {
          return original.bind(target);
        }
        return original;
      }
    });
  }
  beginTransaction() {
    throw new VercelPostgresKyselyError(
      "kysely_transactions_not_supported",
      "Transactions are not supported yet."
    );
  }
  commitTransaction() {
    throw new VercelPostgresKyselyError(
      "kysely_transactions_not_supported",
      "Transactions are not supported yet."
    );
  }
  rollbackTransaction() {
    throw new VercelPostgresKyselyError(
      "kysely_transactions_not_supported",
      "Transactions are not supported yet."
    );
  }
};
function createKysely(config) {
  return new Kysely({
    dialect: new VercelPostgresDialect({
      ...config,
      pool: createPool(config)
    })
  });
}
export {
  createKysely
};
